### PR TITLE
rh-che #557: Adding property and handler for stopping k8s / openshift workspace if unrecoverable event occurs during workspace startup

### PR DESCRIPTION
**Pull request by ibuziuk**
### What does this PR do?
Adding property and handler for stopping k8s / openshift workspace if unrecoverable event occurs during workspace startup

Demo - https://youtu.be/J_DfNZXidE0

### What issues does this PR fix or reference?
https://github.com/redhat-developer/rh-che/issues/557

Adding property `che.infra.kubernetes.workspace_unrecoverable_events`

if during workspace startup an unrecoverable event occurs (e.g. "Failed Mount" "Failed Scheduling") terminate workspace immediately instead of waiting util timeout

#### Release Notes
Adding property `che.infra.kubernetes.workspace_unrecoverable_events`

if during workspace startup an unrecoverable event occurs (e.g. "Failed Mount" "Failed Scheduling") terminate workspace immediately instead of waiting util timeout

#### Docs PR
https://github.com/eclipse/che-docs/pull/395

### CODE REVIEW

File: infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntimeTest.java [hunk 1]
```diff
@@ -193,6 +195,7 @@ public void shouldStartMachines() throws Exception {
     verify(routes).create(any());
     verify(services).create(any());
 
+    verify(project.pods(), times(1)).watchContainers(any());
```
Comment (garagatyi) [inline infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntimeTest.java]: Can you also check for the cancellation of watching?

File: infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntimeTest.java [hunk 2]
```diff
@@ -137,6 +138,7 @@ public void setup() throws Exception {
         new OpenShiftInternalRuntime(
             13,
             5,
+            FAIL_ON_UNRECOVERABLE_EVENT,
```
Comment (garagatyi) [inline infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntimeTest.java]: Can you add a case when the value is false?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java [hunk 3]
```diff
@@ -420,6 +426,10 @@ protected void startMachines() throws InfrastructureException {
     // namespace.pods().watch(new AbnormalStopHandler());
     // namespace.pods().watchContainers(new MachineLogsPublisher());
 
+    if (failFastOnUnrecoverableEvents) {
+      namespace.pods().watchContainers(new UnrecoverableEventHanler());
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: I think it could be useful add `MachineLogsPublisher` here. WDYT?

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntimeTest.java [hunk 4]
```diff
@@ -180,6 +182,7 @@ public void setup() throws Exception {
         new KubernetesInternalRuntime<>(
             13,
             5,
+            FAIL_ON_UNRECOVERABLE_EVENT,
```
Comment (garagatyi) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntimeTest.java]: Can you add a case when the value is false?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java [hunk 5]
```diff
@@ -600,6 +610,26 @@ public void accept(ProbeResult probeResult) {
...
+      if (isUnrecoverable(reason)) {
+        LOG.error("Unrecoverable event occured {}, {}", reason, message);
+        eventPublisher.sendRuntimeStoppedEvent(
+            format("Unrecoverable event occured: '%s', '%s'", reason, message),
+            getContext().getIdentity());
+      }
+    }
+
+    private boolean isUnrecoverable(String reason) {
+      return reason.equals("Failed Mount") || reason.equals("Failed Scheduling");
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: Maybe it makes sense to make these values configurable?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java [hunk 6]
```diff
@@ -600,6 +610,26 @@ public void accept(ProbeResult probeResult) {
     }
   }
 
+  /** Listens container's events and terminates runtime if unrecoverable event occurs. */
```
Comment (garagatyi) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: AFAIK we don't have runtime interruption in place for k8s runtime yet. Can you elaborate how this handler stops start of a runtime?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java (hunk 5 above)
Comment (garagatyi) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: +1

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java (hunk 3 above)
Comment (garagatyi) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: @sleshchenko can you elaborate on what exactly to do?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java (hunk 5 above)
Comment (ibuziuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: @sleshchenko do you mean property with comma separated reasons ?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/event/ContainerEvent.java [hunk 7]
```diff
@@ -80,6 +88,9 @@ public String toString() {
         + ", containerName='"
         + containerName
         + '\''
+        + ", reason='"
```
[the remark this answers is not in the document]
Comment (ibuziuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/event/ContainerEvent.java:92]: yeah, good point

File: infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntimeTest.java (hunk 2 above)
Comment (ibuziuk) [reply infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntimeTest.java]: @garagatyi do you think it is ok to check one infra with `false` & other with `true`, or you think there should be both cases handled in both infras ?

Comment (garagatyi) [reply infrastructures/openshift/src/test/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntimeTest.java]: Since they might have their own bugs we should have tests for both

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java [hunk 8]
```diff
@@ -600,6 +610,26 @@ public void accept(ProbeResult probeResult) {
...
+  public class UnrecoverableEventHanler implements ContainerEventHandler {
+
+    @Override
+    public void handle(ContainerEvent event) {
+      String reason = event.getReason();
+      String message = event.getMessage();
+      if (isUnrecoverable(reason)) {
+        LOG.error("Unrecoverable event occured {}, {}", reason, message);
```
[the remark this answers is not in the document]
Comment (garagatyi) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: Maybe workspace ID, maybe some pod/container info if any

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java [hunk 9]
```diff
@@ -240,6 +244,8 @@ private void waitMachines(
       failure.completeExceptionally(ex);
       cancelAll(toCancelFutures);
       wrapAndRethrow(ex.getCause());
+    } finally {
+      namespace.pods().stopWatch();
```
[the remark this answers is not in the document]
Comment (akorneta) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: +1

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java (hunk 5 above)
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: I think yes.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java (hunk 3 above)
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: I think it should be something like the following:
```
--> start
try {
  namespace.pods().watchContainers(new MachineLogsPublisher());
  if (failFastOnUnrecoverableEvents) {
    namespace.pods().watchContainers(new UnrecoverableEventHanler());
  }

  //doStart here
} catch ....
... 
} finally {
  namespace.pods().stopWatch();
}
```
In this case clients would receive K8s/OS events text as machine's output, and it would be easier to understand why workspace start fails.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/event/ContainerEvent.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (ibuziuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/event/ContainerEvent.java:92]: fixed

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java (hunk 3 above)
Comment (ibuziuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: @sleshchenko Adding `MachineLogsPublisher` seems to be unrelated to this issue but rather to https://github.com/eclipse/che/issues/7653 IMO. Could this be addressed in another PR ?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (ibuziuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: have added pod info

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java (hunk 6 above)
Comment (ibuziuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: s/runtime/workspace/

**Review by akorneta — APPROVED** [review]

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java (hunk 3 above)
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: @ibuziuk For me, it looks like it's easy to add it here (I guess described changes should be enough) and it will be useful (machine logs panel will have all events). It's why I write it here. 
But I agree that it is not related to a solved issue and it can be added in another PR.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (ibuziuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: sounds good, let me fix

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java (hunk 3 above)
Comment (ibuziuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: @sleshchenko ok, sounds relevant - let me add it to this PR

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java (hunk 5 above)
Comment (ibuziuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: done

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (ibuziuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: moved to `internalStart`

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 10]
```diff
@@ -347,6 +347,11 @@ che.infra.kubernetes.workspace_start_timeout_min=8
 # Defines the timeout in minutes that limits the period for which Kubernetes Ingress become ready
 che.infra.kubernetes.ingress_start_timeout_min=5
 
+# If during workspace startup an unrecoverable event defined in 'che.infra.kubernetes.workspace_unrecoverable_events' occurs
+# terminate workspace immediately instead of waiting util timeout
```
Comment (sleshchenko) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: Should `util` be `until`?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java [hunk 11]
```diff
@@ -600,8 +612,35 @@ public void accept(ProbeResult probeResult) {
...
+  public class UnrecoverableEventHanler implements ContainerEventHandler {
+
+    @Override
+    public void handle(ContainerEvent event) {
+      if (isUnrecoverable(event)) {
+        String reason = event.getReason();
+        String message = event.getMessage();
+        String podName = event.getPodName();
+        try {
+          internalStop(emptyMap());
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java:632]: I think it doesn't work correctly until k8s runtime start interruption will be implemented. 
I guess client will receive two workspace status events:
- one that workspace failed to start because of Unrecoverable;
- a second one that start timeout exceeded.

Let me check that. BTW I think we can merge these changes with the described issue and it will be fixed with implemented interruption of k8s runtime start.

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 12]
```diff
@@ -347,6 +347,11 @@ che.infra.kubernetes.workspace_start_timeout_min=8
...
+che.infra.kubernetes.workspace_fail_on_unrecoverable_events=true
```
Comment (sleshchenko) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: Can you describe why it is necessary to have separate configuration property for disabling listening to unrecoverable events. Maybe it makes sense to configure empty `che.infra.kubernetes.workspace_unrecoverable_events` for disabling instead of having a separate one?

**Review by sleshchenko — APPROVED** [review]
Your changes will make life easier during an investigation of workspace start failing :+1:

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties (hunk 10 above)
Comment (ibuziuk) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: good catch

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties (hunk 12 above)
Comment (ibuziuk) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: good point, less config we have the better

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java (hunk 11 above)
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java:632]: I've checked. So here is the Che Server logs
```

  | 2018-04-17 08:02:45,912[io-8080-exec-10]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 236]   - Starting workspace 'che/qwe' with id 'workspacea75oxtgryzdjq5iq' by user 'che'
  /////////////////// Here Client received that workspace failed to start but Che Server doesn't contain the corresponding message ///////////////////////////
  /////////////////// ~3 minutes later reached workspace start timeout ///////////////////////////
  | 2018-04-17 08:05:18,006[aceSharedPool-0]  [WARN ] [.i.k.KubernetesInternalRuntime 177]  - Failed to start Kubernetes runtime of workspace workspacelhu15ifhded6jrxf. Cause: Server 'terminal' in machine 'dev-machine' not available.
  | 2018-04-17 08:05:18,006[ServersChecker]   [ERROR] [.i.k.KubernetesInternalRuntime 335]  - Unable to update status of the machine '%s:%s'. Cause: %s
  | 2018-04-17 08:05:18,066[aceSharedPool-0]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 292]   - Workspace 'che:wksp-g9f7' with id 'workspacelhu15ifhded6jrxf' start failed
  | 2018-04-17 08:05:18,070[aceSharedPool-0]  [WARN ] [o.e.c.a.w.s.WorkspaceManager 423]    - Cannot set error status of the workspace workspacelhu15ifhded6jrxf. Error is: Workspace with id 'workspacelhu15ifhded6jrxf' doesn't exist
```

File: infrastructures/kubernetes/pom.xml [hunk 13]
```diff
@@ -66,6 +66,10 @@
             <groupId>javax.inject</groupId>
             <artifactId>javax.inject</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.codehaus.plexus</groupId>
```
Comment (skabashnyuk) [inline infrastructures/kubernetes/pom.xml]: I guess we don't have CQ for this dependency. Can you use some alternatives?

Comment (ibuziuk) [reply infrastructures/kubernetes/pom.xml]: @skabashnyuk seems to be defined in parent https://github.com/eclipse/che-parent/blob/master/pom.xml#L252

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 14]
```diff
@@ -347,6 +347,10 @@ che.infra.kubernetes.workspace_start_timeout_min=8
 # Defines the timeout in minutes that limits the period for which Kubernetes Ingress become ready
 che.infra.kubernetes.ingress_start_timeout_min=5
 
+# If during workspace startup an unrecoverable event defined in the property occurs,
+# terminate workspace immediately instead of waiting until timeout
+che.infra.kubernetes.workspace_unrecoverable_events=Failed Mount,Failed Scheduling,Failed to pull image
```
Comment (l0rd) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:358]: For users it may be interesting to have a list (or better a link) of valid kubernetes events.

**Review by l0rd — APPROVED** [review]

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties (hunk 14 above)
Comment (ibuziuk) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:358]: @l0rd could you please clarify a bit ? is it about changing the property format

File: infrastructures/kubernetes/pom.xml (hunk 13 above)
Comment (skabashnyuk) [reply infrastructures/kubernetes/pom.xml]: It's not an issue where it's defined. the problem is that it becomes compile level and we have no cq for it. https://dev.eclipse.org/ipzilla/buglist.cgi?query_format=advanced&short_desc_type=allwordssubstr&short_desc=&component=ecd.che&long_desc_type=substring&long_desc=&bug_file_loc_type=allwordssubstr&bug_file_loc=&keywords_type=allwords&keywords=&emailassigned_to1=1&emailtype1=substring&email1=&emailassigned_to2=1&emailreporter2=1&emailcc2=1&emailtype2=substring&email2=&bugidtype=include&bug_id=&chfieldfrom=&chfieldto=Now&chfieldvalue=&cmdtype=doit&order=Reuse+same+sort+as+last+time&field0-0-0=noop&type0-0-0=noop&value0-0-0=

AFAIK you can find the same helper method on other artifacts like guava

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties (hunk 14 above)
Comment (garagatyi) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:358]: I think Mario means a link to docs where all the event types are listed. 
And :+1:  to add it here.

Comment (l0rd) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:358]: If I am a Che user and I am editing che.properties I don't know what are the valid values for this new property. You may want to say in the comment that this is a comma separated list of kubernetes events. And it would be great if we could add a link to a list of valid kubernetes events (but I am not sure we can easily find one so don't lose too much time on this)

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java [hunk 15]
```diff
@@ -600,8 +611,42 @@ public void accept(ProbeResult probeResult) {
...
+        } catch (InfrastructureException e) {
+          LOG.error("Unrecoverable event occured {}, {}, {}", reason, message, podName);
```
Comment (garagatyi) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: It seems that without workspace ID it won't be possible to match the incident with a workspace/user. I think we need WSID her.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/event/ContainerEvent.java [hunk 16]
```diff
@@ -62,6 +70,7 @@ public boolean equals(Object o) {
     ContainerEvent that = (ContainerEvent) o;
     return Objects.equals(podName, that.podName)
         && Objects.equals(containerName, that.containerName)
+        && Objects.equals(reason, that.reason)
```
Comment (garagatyi) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/event/ContainerEvent.java:73]: Can you also add the reason to the hashcode method? It is common to have them aligned.

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java [hunk 17]
```diff
@@ -92,7 +97,11 @@ protected void startMachines() throws InfrastructureException {
     }
     // TODO https://github.com/eclipse/che/issues/7653
     // project.pods().watch(new AbnormalStopHandler());
-    // project.pods().watchContainers(new MachineLogsPublisher());
+
+    project.pods().watchContainers(new MachineLogsPublisher());
```
Comment (garagatyi) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java:101]: TLDR please, add finally here as well
Since we set watchers here we should also ensure that all the watchers are properly deleted. We can not guarantee that there is no bug which prevents the execution of finally block from k8s infra.

**Comment by garagatyi** [comment]
@ibuziuk please, do not push to the branch with force. Each force push make reviewers review whole PR from scratch

**Comment by garagatyi** [comment]
BTW looks like you need to address @skabashnyuk 's comment before merging since we can't add new dependency without a CQ

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java (hunk 17 above)
Comment (ibuziuk) [reply infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java:101]: sorry, not following - `startMachines` is supposed to be called from `internalStart` where watchers closed / deleted in finally. Could you describe potential bug that could prevent closing watchers currently ?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java (hunk 15 above)
Comment (ibuziuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java]: ok, will also add workspace id

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/event/ContainerEvent.java (hunk 16 above)
Comment (ibuziuk) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/event/ContainerEvent.java:73]: done, thanks for the pointer

File: infrastructures/kubernetes/pom.xml (hunk 13 above)
Comment (ibuziuk) [reply infrastructures/kubernetes/pom.xml]: @skabashnyuk ok, good to know - changed to `com.google.common.base.Strings`

**Comment by ibuziuk** [comment]
@garagatyi fixed dep CQ problem -  changed to `com.google.common.base.Strings`

**Comment by ibuziuk** [comment]
@l0rd have added PR for docs https://github.com/eclipse/che-docs/pull/395

**Comment by garagatyi** [comment]
@ibuziuk is there any other cases that you are going to cover in unit tests?

**Comment by ibuziuk** [comment]
@garagatyi yeah, working on adding more tests

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 18]
```diff
@@ -347,6 +347,10 @@ che.infra.kubernetes.workspace_start_timeout_min=8
 # Defines the timeout in minutes that limits the period for which Kubernetes Ingress become ready
 che.infra.kubernetes.ingress_start_timeout_min=5
 
+# If during workspace startup an unrecoverable event defined in the property occurs,
+# terminate workspace immediately instead of waiting until timeout
+che.infra.kubernetes.workspace_unrecoverable_events=Failed Mount,Failed Scheduling,Failed
```
Comment (l0rd) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: `Failed to pull image` has become `Failed`

**Review by l0rd — CHANGES_REQUESTED** [review]

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties (hunk 18 above)
Comment (ibuziuk) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: it was actually done deliberate, but I can rework

Comment (l0rd) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties]: Yes please revert that. Having `Failed` in a list of kubernetes events is confusing.

**Review by skabashnyuk — APPROVED** [review]

**Comment by ibuziuk** [comment]
@l0rd @garagatyi changed property value, added tests. Please, review

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java (hunk 17 above)
Comment (garagatyi) [reply infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/OpenShiftInternalRuntime.java:101]: Sorry, my bad. We don't need finally here.

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntimeTest.java [hunk 19]
```diff
@@ -352,8 +412,9 @@ public void throwsInfrastructureExceptionWhenKubernetesNamespaceCleanupFailed()
   @Test
   public void testRepublishContainerOutputAsMachineLogEvents() throws Exception {
     final MachineLogsPublisher logsPublisher = internalRuntime.new MachineLogsPublisher();
-    final ContainerEvent out1 = mockContainerEvent("pulling image", "07/07/2007 19:01:22");
-    final ContainerEvent out2 = mockContainerEvent("image pulled", "07/07/2007 19:08:53");
+    final ContainerEvent out1 =
```
Comment (garagatyi) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntimeTest.java:415]: You have added logs publishing in the runtime start process and this is very cool. Can you also add a unit test for this case?

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntimeTest.java [hunk 20]
```diff
@@ -366,10 +427,34 @@ public void testRepublishContainerOutputAsMachineLogEvents() throws Exception {
     assertTrue(captor.getAllValues().containsAll(machineLogs));
   }
 
+  @Test
+  public void testHandleUnrecoverableEvent() throws Exception {
+    final UnrecoverableEventHanler unrecoverableEventHanler =
```
Comment (garagatyi) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntimeTest.java]: Please, test unrecoverable events matching/non-matching in the whole flow instead of with a separated component. We don't have any guarantees that handler is called as it is supposed in this test case.
Please cover cases:
non-matching event
matching by the reason field
matching by the message field

It would be nice to verify that stop of the runtime is really called after appearing of the matching event.

**Review by l0rd — APPROVED** [review]
LGTM. Thanks @ibuziuk!

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntimeTest.java (hunk 19 above)
Comment (ibuziuk) [reply infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntimeTest.java:415]: isn't it already handled by https://github.com/eclipse/che/pull/9426/files#diff-08f3ea6fea2790aa766edf8d57d37897R275 e.g. handler for publishing events is added during runtime start?

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntimeTest.java (hunk 20 above)
Comment (ibuziuk) [reply infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntimeTest.java]: @garagatyi was not able to figure out how exactly to test unrecoverable events in starting flow, but have added separate tests for matching by the reason / matching by the message. Also, not sure how exactly verifying that stop of the runtime e.g. `internalStop` should be tested - currently done by verifying that namespace cleanup was called

Are you ok to merge it in the current state, since I do not really want to face merge conflicts, and we could discuss more tests cases which I would add later in separate PR ?

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntimeTest.java (hunk 19 above)
Comment (garagatyi) [reply infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntimeTest.java:415]: Well, it just checks that some subscribers are watching it. We have a separate test for logs publisher, so maybe we can also check that one of the watchers is logs publisher and another one is unrecoverable events listener? But it is not critical, so feel free to leave it as is if you want.

**Review by garagatyi — APPROVED** [review]
Looks good! I believe that tests that we were discussing are not mandatory for the PR.

**Comment by ibuziuk** [comment]
@garagatyi ok, thanks I will try to add more tests for those event handlers during the sprint
